plugins/in_winevtlog/winevtlog.c

```diff
@@ -0,0 +1,792 @@
+#include <fluent-bit/flb_input.h>
+#include "winevtlog.h"
+
+#define EVENT_PROVIDER_NAME_LENGTH 256
+
+static char* convert_wstr(wchar_t *wstr, UINT codePage);
+static wchar_t* convert_str(char *str);
+
+struct winevtlog_channel *winevtlog_subscribe(const char *channel, int read_existing_events,
+                                              EVT_HANDLE stored_bookmark)
+{
+    struct winevtlog_channel *ch;
+    EVT_HANDLE bookmark = NULL;
+    HANDLE signal_event = NULL;
+    DWORD len, flags = 0L;
+    PWSTR wide_channel = L"Application", wide_query = L"*";
+    void *buf;
+
+    ch = flb_calloc(1, sizeof(struct winevtlog_channel));
+    if (ch == NULL) {
+        flb_errno();
+        return NULL;
+    }
+
+    ch->name = flb_strdup(channel);
+    if (!ch->name) {
+        flb_errno();
+        flb_free(ch);
+        return NULL;
+    }
+
+    signal_event = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    // channel : To wide char
+    len = MultiByteToWideChar(CP_UTF8, 0, channel, -1, NULL, 0);
+    wide_channel = flb_malloc(sizeof(PWSTR) * len);
+    MultiByteToWideChar(CP_UTF8, 0, channel, -1, wide_channel, len);
+
+    if (stored_bookmark) {
+        flags |= EvtSubscribeStartAfterBookmark;
+    } else if (read_existing_events) {
+        flags |= EvtSubscribeStartAtOldestRecord;
+    } else {
+        flags |= EvtSubscribeToFutureEvents;
+    }
+
+    ch->subscription = EvtSubscribe(NULL, signal_event, wide_channel, wide_query,
+                                    stored_bookmark, NULL, NULL, flags);
+    if (!ch->subscription) {
+        flb_error("[in_winevtlog] cannot subscribe '%s' (%i)", channel, GetLastError());
+        flb_free(ch->name);
+        flb_free(ch);
+        return NULL;
+    }
+    ch->signal_event = signal_event;
+
+    if (stored_bookmark) {
+        ch->bookmark = stored_bookmark;
+    }
+    else {
+        bookmark = EvtCreateBookmark(NULL);
+        if (bookmark) {
+            ch->bookmark = bookmark;
+        }
+        else {
+            if (ch->subscription) {
+                EvtClose(ch->subscription);
+            }
+            if (signal_event) {
+                CloseHandle(signal_event);
+            }
+            flb_error("[in_winevtlog] cannot subscribe '%s' (%i)", channel, GetLastError());
+            flb_free(wide_channel);
+            flb_free(ch->name);
+            flb_free(ch);
+            return NULL;
+        }
+    }
+
+    flb_free(wide_channel);
+
+    return ch;
+}
+
+BOOL cancel_subscription(struct winevtlog_channel *ch)
+{
+    return EvtCancel(ch->subscription);
+}
+
+static void close_handles(struct winevtlog_channel *ch)
+{
+    if (ch->subscription) {
+        EvtClose(ch->subscription);
+        ch->subscription = NULL;
+    }
+    if (ch->signal_event) {
+        CloseHandle(ch->signal_event);
+        ch->signal_event = NULL;
+    }
+    if (ch->bookmark) {
+        EvtClose(ch->bookmark);
+        ch->bookmark = NULL;
+    }
+    for (int i = 0; i < ch->count; i++) {
```

Comment:
Please move this definition to the top

Reply:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/f18098bc507d377ad66b547db1e984929fd66a75.